### DISCUSSION TITLE
Allow decal textures to be scaled / have some sort of UV

### DISCUSSION THREAD

**Discussion by FireCatMagic** [godotengine/godot-proposals#7103] [3D]
I'd like to use a repeating texture as a decal while using the smaller sized texture, but this currently is not possible.
The workaround that I use is to manually resize the texture on runtime, but that can lead to overly large textures.

**Comment by Calinou** [comment]
Adding UV offset/scale properties is an interesting idea, but it makes the decal shader more complex. This means this may impact performance even if you leave these at the default values.

Could you describe the concrete use case(s) you need this for?

**Reply by FireCatMagic** [reply]
![Godot_v4 1-beta2_win64_iagGRHpB6c](https://github.com/godotengine/godot-proposals/assets/40964335/e5449852-0e37-48ef-858f-07d7e1db2b07)

In my game, I use a decal to simulate a water caustics system. It looks alright in my opinion, but I don't think my implementation is good for memory or process time. What it does is when the water node is scaled, it'll generate an entirely new noise texture to fit the new scale. I had to clamp it because when the water was too big, it would generate a texture that's way too big and make the game either super laggy or freeze entirely. For the sin waves of the caustics, I manually adjusted the offset of the noise texture itself in gdscript every frame.

Having this feature would allow me to reduce all of this to simply generating a uv scale based on the scale of the water, allowing me to make the size of my water node infinite. I could then adjust the offset for the sin waves instead of the direct texture.

**Reply by Calinou** [reply]
On the other hand, it probably sounds like we should have support for projectors in DirectionalLight3D instead (using a tiling texture). Using decals for caustics or cloud shadows is an inaccurate approximation, as it'll also affect areas in the shade and will cause brightened areas to lose saturation (as seen on the GIF you posted).

In your particular case, you could toggle the projector texture when the camera enters the water (and set a `null` texture when the camera leaves it).
